var/spack/repos/builtin/packages/unifyfs/package.py

```diff
@@ -16,55 +16,50 @@ class Unifyfs(AutotoolsPackage):
 
     homepage = "https://github.com/LLNL/UnifyFS"
     git      = "https://github.com/LLNL/UnifyFS.git"
-    url      = "https://github.com/LLNL/UnifyFS/releases/download/v0.9.0/unifyfs-0.9.0.tar.gz"
+    url      = "https://github.com/LLNL/UnifyFS/releases/download/v0.9.2/unifyfs-0.9.2.tar.gz"
     maintainers = ['CamStan']
 
     version('develop', branch='dev')
+    version('0.9.2', sha256='7046625dc0677535f5d960187cb2e2d58a6f8cfb4dc6a3604f825257eb0891aa')
     version('0.9.1', sha256='2498a859cfa4961356fdf5c4c17e3afc3de7e034ad013b8c7145a622ef6199a0')
-    version('0.9.0', sha256='e6c73e22ef1c23f3141646aa17058b69c1c4e526886771f8fe982da924265b0f')
+    version('0.9.0', sha256='e6c73e22ef1c23f3141646aa17058b69c1c4e526886771f8fe982da924265b0f', deprecated=True)
 
     variant('auto-mount', default='True', description='Enable automatic mount/unmount in MPI_Init/Finalize')
     variant('hdf5', default='False', description='Build with parallel HDF5 (install with `^hdf5~mpi` for serial)')
     variant('fortran', default='False', description='Build with gfortran support')
-    variant('mdhim', default='False', description='Enable MDHIM build options')
     variant('pmi', default='False', description='Enable PMI2 build options')
     variant('pmix', default='False', description='Enable PMIx build options')
     variant('spath', default='True', description='Use spath library to normalize relative paths')
 
-    depends_on('autoconf',  type='build')
-    depends_on('automake',  type='build')
-    depends_on('libtool',   type='build')
-    depends_on('m4',        type='build')
-    depends_on('pkgconfig', type='build')
+    depends_on('autoconf',        type='build')
+    depends_on('automake',        type='build')
+    depends_on('automake@1.15:',  type='build', when='@0.9.2:')
+    depends_on('libtool',         type='build')
+    depends_on('m4',              type='build')
+    depends_on('pkgconfig',       type='build')
 
     # Required dependencies
-    depends_on('gotcha@1.0.3:', when='@0.9.1:')
-    depends_on('mochi-margo@0.4.3')
-    depends_on('mercury@1.0.1+bmi+sm')
+    depends_on('gotcha@1.0.3:')
+    depends_on('mercury@1.0.1+bmi', when='@:0.9.1')
+    depends_on('mochi-margo@0.4.3', when='@:0.9.1')
+    depends_on('mochi-margo',       when='@0.9.2:')
     depends_on('mpi')
     depends_on('openssl')
 
     # Optional dependencies
     depends_on('hdf5', when='+hdf5')
-    depends_on('leveldb', when='@0.9.1:+mdhim')
-    depends_on('spath', when='@0.9.1:+spath')
+    depends_on('libfabric fabrics=rxm,sockets,tcp', when="^mercury@2:+ofi")
```

Comment:
(question) these constraints look quite specific. Do you happen to know if this package (`unifyfs`) detects the presence of `mercury` and then uses specific features from `libfabric` if so?

FWIW `mercury` has its own dependency on `libfabric`, so unless `unifyfs` is accessing symbols in `libfabric` directly, it may not need this.

Reply:
`mercury` does have a dependency of `libfabric`, however, `mercury`'s spackage doesn't specify the fabrics, and therefore gets the default of `fabrics=sockets,tcp,udp`.

In `unifyfs`, we default to Mercury v2.0 'ofi+tcp' which requires `libfabric fabrics=rxm,tcp` (see the discussion with the Mochi team here: https://github.com/LLNL/UnifyFS/issues/584#issuecomment-772023745), and then fallback to 'ofi+sockets' if the former doesn't work.

Thus the spec of `^libfabric fabrics=rxm,sockets,tcp` allows `unifyfs` to work on more machines.

I may have oversimplified this or missed something. @michaelbrim could clarify more if need be.

Reply:
Possibly also relevant:
Earlier in the above linked issue/comment, one of our other developers stated:
> Mercury v1 converted 'ofi+tcp' to sockets, whereas Mercury v2 now converts it to 'tcp;ofi+rxm'.

So the default fabrics likely still works with Mercury v1.

Reply:
> In unifyfs, we default to Mercury v2.0 'ofi+tcp' which requires libfabric fabrics=rxm,tcp

Does that mean `mercury` itself requires that? If so, I think it would make sense to add this as a constraint to the `mercury` package, i.e.:

```
Mercury(CMakePackage):
    ...
    depends_on(libfabric fabrics=rxm,tcp, when='+ofi+tcp')
```

Reply:
My apologies, I think I created some syntax confusion with how I worded my comment.

When using Spack and we have the dependency of `^mercury+ofi`, this is simply telling Mercury to use the OFI libfabric plugin.
There isn't an option for `+tcp` here.

When using the OFI plugin, there are then several protocols/lookup formats available for initializing Mercury (tcp, verbs, psm2, cmi, sockets). Using "tcp" here is what determines the need for `fabrics=rxm,tcp`. However, the user could provide other combinations that change the fabrics needed.

I'm assuming the Mercury package doesn't specify a certain set of fabrics for OFI then as what the user actually needs will be determined by which protocol(s) they initialize Mercury with and can't be determined at build time.
I suppose they could provide variants for all the combinations so that you know you built with the right fabrics depending on the protocols you intend to use, but I imagine that might get complicated rather quickly.

---

For example:
In the UnifyFS code, when initializing Mercury (which we do though Margo) we tell it which protocol/lookup format to use. This syntax is similar to Spacks, but without the initial `+`:
``` C
margo_instance_id mid;
const char* margo_protocol;

margo_protocol = "ofi+tcp";
mid = margo_init(margo_protocol, ...additional args...);
```

I apologize if any of this is confusing as this is outside of the scope of what I've been working on with UnifyFS. @michaelbrim, please correct me if I'm wrong here.

Reply:
Thanks for the explanation! It looks like you wrote the constraint exactly as needed. I hope it wasn't too annoying but I was not familiar with the packages and was curious how they interacted.

Reply:
Not a problem at all. I'm often not sure what to do in Spack packages when something involves a dependency of a dependency. If nothing else, it was good for me to think through this.